### PR TITLE
std.range: replace some enforces with contracts

### PR DESCRIPTION
**Pull request by John-Colvin**
These changes are in keeping with the vision of ```assert```s being for program bugs.

#### Advantages: ####
* std.range.stride is now usable from ```@nogc``` and ```nothrow```.
* std.range.iota for non-floating types now works with ```@nogc``` and ```nothrow```.

This is a breaking change for anyone who relied on the old exceptions to validate their input.

### CODE REVIEW

File: std/range/package.d [hunk 1]
```diff
@@ -4196,6 +4209,12 @@ auto iota(E)(E end)
 // Specialization for floating-point types
 auto iota(B, E, S)(B begin, E end, S step)
 if (isFloatingPoint!(CommonType!(B, E, S)))
+in
+{
+    assert(step != 0, "iota: step must not be 0");
+    assert((end - begin) / step >= 0, "iota: incorrect startup parameters");
```
Comment (JakobOvrum) [inline std/range/package.d]: AFAICS, type `S` can easily be a floating-point number, which could make these equality comparisons flaky. How close to 0 is illegal, and is it really OK that different floating-point types for `S` change the margin? Does comparing to 0 only check the two special values +0.0 and -0.0? Or is it just +0.0?

Comment (John-Colvin) [inline std/range/package.d]: I just took them from (old) lines 4213 and 4215.

iota needs a little t.l.c. w.r.t. numerical concerns, but that's outside the scope of this pull.

File: std/range/package.d [hunk 2]
```diff
@@ -404,18 +404,18 @@ Iterates range $(D r) with stride $(D n). If the range is a
 random-access range, moves by indexing into the range; otherwise,
 moves by successive calls to $(D popFront). Applying stride twice to
 the same range results in a stride with a step that is the
-product of the two applications.
-
-Throws: $(D Exception) if $(D n == 0).
+product of the two applications. It is an error for $(D n) to be 0.
```
Comment (monarchdodra) [inline std/range/package.d]: Not a huge fan of such wording. I'd have expected a "$(D n) must not be empty" or "$(D n) shall not be empty". But whatevs.

**Comment by monarchdodra** [comment]
LGTM. Great initiative.

**Comment by monarchdodra** [comment]
Auto-merge toggled on

**Comment by andralex** [comment]
This is a subtle change in semantics that went over with rather little scrutiny. I understand the argument about `@nogc` but `nothrow` was actually a natural consequence of argument checking. There should be an organized discussion about parameter validation in Phobos; this dribble of small diffs is not the right approach.

**Comment by John-Colvin** [comment]
I agree. I was expecting some response to my "this is a breaking change"
sentence.

Having said that I don't think that std.algorithm and std.range have any
business doing input validation. These are the primitive iteration schemes
and algorithms at the core of D, they have to move at full speed or it
seriously limits their usefulness.

**Comment by monarchdodra** [comment]
> This is a subtle change in semantics that went over with rather little scrutiny

> I agree. I was expecting some response to my "this is a breaking change"
sentence.

We have gradually but consistently changing these exceptions to asserts in every range/algorithm/container function for quite some time now (like 2 years). 1 commit at a time, they get systematically merged in with no objection whatsoever.

I'm all fine with starting a discussion, but quite frankly, I don't even know what there is to discuss. Do note that in release (AFAIK), memory bounds checking *are* still in operation where needed, so the code is still safe (which is actually tricky to get right with asserts and manual allocations).

**Pull request merged**